test/cpu_multimarginloss.hpp

```diff
@@ -0,0 +1,112 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2024 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#ifndef GUARD_CPU_MULTIMARGINLOSS_HPP
+#define GUARD_CPU_MULTIMARGINLOSS_HPP
+
+#include "tensor_holder.hpp"
+#include <miopen/tensor_view_utils.hpp>
+
+template <class T>
+void cpu_multimarginloss_unreduced_forward(tensor<T> input,
+                                           tensor<uint64_t> target,
+                                           tensor<T> weight,
+                                           tensor<T>& ref_output,
+                                           long p,
+                                           float margin)
+{
+    auto I_tv = miopen::get_inner_expanded_tv<2>(input.desc);
+    auto T_tv = miopen::get_inner_expanded_tv<1>(target.desc);
+    auto W_tv = miopen::get_inner_expanded_tv<1>(weight.desc);
+    auto O_tv = miopen::get_inner_expanded_tv<1>(ref_output.desc);
+    auto N = I_tv.size[0], C = I_tv.size[1];
+
+    for(size_t n = 0; n < N; n++)
+    {
+        float loss = 0;
```

Comment:
Using `double` to reduce floating point precision.

Reply:
And please do the same to the rest of this file.

Reply:
fixed it